kyo-scheduler/jvm/src/main/scala/kyo/scheduler/regulator/Regulator.scala

```diff
@@ -0,0 +1,96 @@
+package kyo.scheduler.regulator
+
+import Regulator.*
+import kyo.scheduler.InternalTimer
+import kyo.scheduler.util.*
+import kyo.stats.internal.MetricReceiver
+import org.slf4j.LoggerFactory
+import scala.util.control.NonFatal
+
+abstract class Regulator(
+    loadAvg: () => Double,
+    timer: InternalTimer,
+    config: Config
+):
+    import config.*
+
+    private var step         = 0
+    private val measurements = MovingStdDev(collectWindow)
+
+    protected def probe(): Unit
+    protected def update(diff: Int): Unit
+
+    protected def measure(v: Long): Unit =
+        stats.measurements.observe(v.toDouble)
+        synchronized(measurements.observe(v))
+
+    private val collectTask =
+        timer.schedule(collectInterval)(collect())
+
+    private val regulateTask =
+        timer.schedule(regulateInterval)(adjust())
+
+    final private def collect(): Unit =
+        try
+            if loadAvg() < loadAvgTarget then
+                measure(0)
+            else
+                stats.probes.inc()
+                probe()
+            end if
+        catch
+            case ex if NonFatal(ex) =>
+                log.error(s"🙈 !!Kyo Scheduler Bug!! ${getClass.getSimpleName()} regulator's probe collection has failed.", ex)
```

Comment:
now that we are using a logger instance for this class, we can remove `getClass.getSimpleName()`

Reply:
good catch!

Reply:
ah actually I'm not using Kyo's logging because the idea is for the scheduler module to be isolated without a dependency to kyo-core. This is the regular slf4j logger.

Reply:
Right, but you are using SLF4J instance for each class which should include the classname. See line 92:

```
private[Regulator] val log = LoggerFactory.getLogger(getClass)
```

This goes throughout

Reply:
The `log` in this case would be `Regulator`, not the subclass. But I've decided to remove the `slf4j` dependency since logging is used only for bugs. There's the `LoomSupport` warning but I think it's ok to log that with a regular `println`. The module now has zero dependencies, which should help avoid issues in case it's used in isolation without other Kyo modules.

Reply:
Cool emoji!